docs: basic control flow (#5163)
dfinity/motoko: doc/md/fundamentals/4-control-flow/1-basic-control-flow.md

```motoko no-repl
import Nat "mo:base/Nat";

type HttpRequestStatus = {
  #ok: Nat;
  #err: Nat;
};

func checkStatus(r : HttpRequestStatus) : Text {
  let #ok(status) = r else return "The request failed!";
  Nat.toText(status)
};
```

:::note
Unlike a `switch`, `let-else` discards any additional error information from non-matching cases, making it less suitable when detailed error handling is needed. The `#err(e)` case is dropped entirely — `e` cannot be inspected or logged.
:::

## Option block

These blocks represented as `do ? {...}` allow safe unwrapping of optional values using the postfix operator `!`, which short-circuits and returns `null` if any value is `null`, simplifying code that handles multiple options.

```motoko no-repl
let x : ?Nat = foo(); // x might be null or have a Nat value
let y = ?Nat = foo(); // y might also be null or have a Nat value
let z = do ? { x! + 3 / y! }; // If x or y is null, z is also null; otherwise, z = x + 3 / y
```

Instead of having to switch `x` and `y` in a verbose manner the use of the postfix operator `!` makes it simple to unwrap.

## `label`

> A `label` assigns a name to a block of code that executes like any other block, but its result can be accessed directly. Labels provide more control over execution, allowing clear exit points and helping to structure complex logic effectively.

Reviewer: ```suggestion
A `label` assigns a name to a block of code that executes like any other block, but its result can be produced early using a `break` to the label.
If the block produces a non-`(0)` result, the `break` can include a value.
Labels provide more control over execution, allowing clear exit points and helping to structure complex logic effectively.
```

---
sidebar_position: 1
---

> 

Reviewer: ```suggestion
<!-- TODO(Future): the examples using  `sum` are ok, but artificial. Rewriting them to use  `product` would be more meaningful since you can exit early when a factor is 0 and skip the `*=` when factors are 1 -->
```

# Types of control flow

| Construct | Description |
|--------------|---------------|
| `return` | Exits a function and returns a value. |
| `if` | Executes a block if the condition is `true`. |
| `if/else` | Executes different blocks based on a condition. |
| `switch` | [Pattern matching](https://internetcomputer.org/docs/motoko/fundamentals/pattern-matching) for variants, options, results, etc. |
| `let-else` | Destructures a pattern and handle the failure case inline. |
| `option block` | Evaluates an expression and wraps the result in an option type, allowing scoped handling of `null` values. |
| `label/break` | Allows exiting loops early. |
| `while` | Runs while a condition is `true`. |

> | `for` | Iterates over elements in a collection, terminating when no elements remain. |

Reviewer: Are try-catch and throw discussed elsewhere?

:::

## Option block

These blocks represented as `do ? {...}` allow safe unwrapping of optional values using the postfix operator `!`, which short-circuits and exits the block with `null` if any value is `null`, simplifying code that handles multiple options.
The result of the inner block, if any, is returned in an option.

```motoko no-repl
let x : ?Nat = foo(); // x might be null or have a Nat value
let y : ?Nat = foo(); // y might also be null or have a Nat value
let z = do ? { x! + 3 / y! }; // If x or y is null, z is also null; otherwise, z = ?(x + 3 / y)
```
<!-- TODO(future): better, complete example. Perhaps there's one already the option section -->
Instead of having to switch on the options `x` and `y` in a verbose manner the use of the postfix operator `!` makes it easy to unwrap their values but exit the block with `null` when either is `null`.

## `label`

A `label` assigns a name to a block of code that executes like any other block, but its result can be produced early using a `break` to the label.

> If the block produces a non-`(0)` result, the `break` can include a value.

Reviewer: ```suggestion
If the block produces a non-`()` result, the `break` can include a value.
```

```motoko no-repl
func labelControlFlow() : Int {
  let result = label processNumbers : Int {
    let numbers : [Int] = [3, -1, 0, 5, -2, 7];
    var sum : Int = 0;

    for (number in numbers.values()) {
      sum += number;
    };
    sum; // The final result of the block
  };
  return result;
}
```

### `break` within a labeled block

A `break` statement immediately exits a labeled block and returns a specified value. However, `break` must always refer to a label identifier; it cannot be used without one.

```motoko no-repl
func breakControlFlow() : Int {
  let result = label processNumbers: Int {
    let numbers : [Int] = [3, -1, 0, 5, -2, 7];
    var sum : Int = 0;

    for (num in numbers.values()) {
      if (num < 0) {
        break processNumbers sum; // Exit early with current sum
      };
      sum += num;
    };
    sum // This is returned if no break occurs
  };
  return result;
}
```

## `while`

A `while` loop repeatedly executes a block of code as long as a specified condition evaluates to `true`.

```motoko no-repl
import Debug "mo:base/Debug";
import Nat "mo:base/Nat";

var i = 0;
while (i < 5) {
  Debug.print(Nat.toText(i));
  i += 1;
}
```

## `for`

> A `for` loop iterates over elements in a collection or range, executing a block of code for each element.

Reviewer: ```suggestion
A `for` loop iterates over the elements of an iterator, executing a block of code for each element.
```

``` motoko no-repl
import Debug "mo:base/Debug";
import Nat "mo:base/Nat";

let numbers = [1, 2, 3, 4, 5];
for (num in numbers.vals()) {
    Debug.print(Nat.toText(num));
}
```

## Program flows

> In Motoko, code executes sequentially, running one statement after another. However, certain constructs can alter this flow, such as exiting a block early, skipping iterations in a loop, returning a value from a function, or invoking another function.

Reviewer: ```suggestion
In Motoko, code executes sequentially, evaluating expressions and declarations in order. However, certain constructs can alter this flow, such as exiting a block early, skipping iterations in a loop, returning a value from a function, or invoking another function.
```

### `continue`

> A `continue` statement skips the remainder of the current iteration in a loop and immediately proceeds to the next iteration. Like `break`, `continue` must reference a label and only works within a labeled loop.

Reviewer: ```suggestion
A `continue` expression skips the remainder of the current iteration in a loop and immediately proceeds to the next iteration. Like `break`, `continue` must reference a label and only works within a labeled `while`, `for` or `loop` expression.
```

```motoko no-repl
import Nat "mo:base/Nat";

persistent actor {

  func processNumbers(numbers : [Int]) : Int {
    var sum : Int = 0;
    for (num in numbers.values()) {
      if (num < 0) {
        return sum;
      };
      sum += num;
    };
    return sum;
  };

  shared func functionCallControlFlow() : async Int {
    let numbers : [Int] = [3, 1, 5, -1, -2, 7];
    return processNumbers(numbers); // Function call
  };

}
```

Execution begins in `functionCallControlFlow()`, where the function `processNumbers()` is invoked, transferring control to its logic. Inside `processNumbers()`, the numbers are processed one by one. If a negative number is encountered, a `return` statement immediately halts the function and returns the current sum. Control then flows back to `functionCallControlFlow()`, which receives the result and returns it.

> Function calls temporarily interrupt the normal sequential flow by shifting execution to a separate block of logic. Once the called function completes, control resumes at the point where the call was made.

Reviewer: ```suggestion
Function calls temporarily interrupt the normal sequential flow by shifting execution to a separate block of logic. Once the called function completes, control resumes at the point where the call was made, continuing with its result.
```
